source/simulator/assembly.cc

```diff
@@ -1419,7 +1419,8 @@ namespace aspect
                    * on the direction of the flow. Thus the flow-dependent terms below are
                    * only calculated if the edge is an inflow edge.
                    */
-                  const bool inflow = ((current_u * scratch.face_finite_element_values->normal_vector(q)) < 0.);
+                  const bool inflow = (advection_field.is_temperature()
+                                       || ((current_u * scratch.face_finite_element_values->normal_vector(q)) < 0.));
```

Comment:
I think @yinghe616 is correct here, so long as we don't allow a degenerate diffusion coefficient. This would then also be consistent with the comment in line 1343 above. The hyperbolic nature of the composition equation means it's obviously only sensible to impose Dirichlet on inflow. The non-degenerate (I presume ASPECT never allows zero temperature diffusion?) elliptic temperature equation should just have one of Dirichlet and Neumann imposed at each boundary point, independent of flow direction (and imposing nothing is identical to zero Neumann). 

This code was originally built to handle the case of a degenerate diffusion, my copy-paste-mistake!

Perhaps it would also be better to rename this something other than `inflow`, though I can't currently think of anything much snappier than `on_composition_inflow_boundary_or_advection_field_is_temperature` ;)

Thanks for catching this @yinghe616!